actors/migration/nv15/miner.go

```diff
@@ -3,79 +3,251 @@ package nv15
+			}
+
+			inSectors, err := miner6.LoadSectors(store, inRoot)
+			if err != nil {
+				return cid.Undef, xerrors.Errorf("failed to load inSectors: %w", err)
+			}
+
+			prevOutSectors, err := miner7.LoadSectors(store, prevOutRoot)
+			if err != nil {
+				return cid.Undef, xerrors.Errorf("failed to load prevOutSectors: %w", err)
+			}
+
+			for _, change := range diffs {
+				switch change.Type {
+				case amt.Remove:
+					if err := prevOutSectors.Delete(change.Key); err != nil {
+						return cid.Undef, xerrors.Errorf("failed to delete sector from prevOutSectors: %w", err)
+					}
+				case amt.Add:
+					fallthrough
+				case amt.Modify:
+					sectorNo := abi.SectorNumber(change.Key)
+					info, found, err := inSectors.Get(sectorNo)
+					if err != nil {
+						return cid.Undef, xerrors.Errorf("failed to get sector %d in inSectors: %w", sectorNo, err)
+					}
+
+					if !found {
+						return cid.Undef, xerrors.Errorf("didn't find sector %d in inSectors", sectorNo)
+					}
+
+					if err := prevOutSectors.Set(change.Key, migrateSectorInfo(*info)); err != nil {
+						return cid.Undef, xerrors.Errorf("failed to set migrated sector %d in prevOutSectors", sectorNo)
+					}
+				}
+			}
+
+			outArray = prevOutSectors.Array
+		} else {
+			// first time we're doing this, do all the work
+			outArray, err = adt.MakeEmptyArray(store, miner7.SectorsAmtBitwidth)
+			if err != nil {
+				return cid.Undef, xerrors.Errorf("failed to construct new sectors array: %w", err)
+			}
+
+			var sectorInfo miner6.SectorOnChainInfo
+			if err = inArray.ForEach(&sectorInfo, func(k int64) error {
+				return outArray.Set(uint64(k), migrateSectorInfo(sectorInfo))
+			}); err != nil {
+				return cid.Undef, err
+			}
+		}
+
+		outRoot, err := outArray.Root()
+		if err != nil {
+			return cid.Undef, xerrors.Errorf("error writing new sectors AMT: %w", err)
+		}
+
+		_ = cache.Write(MinerPrevSectorsInKey(minerAddr), inRoot)
+
+		_ = cache.Write(MinerPrevSectorsOutKey(minerAddr), outRoot)
+		return outRoot, nil
+	})
+}
+
+func (m minerMigrator) migrateDeadlines(ctx context.Context, store adt.Store, deadlines cid.Cid, sectors cid.Cid) (cid.Cid, error) {
+	if deadlines == m.emptyDeadlinesV6 && sectors == m.emptySectorsV7 {
+		return m.emptyDeadlinesV7, nil
 	}
 
-	outArray, err := adt.MakeEmptyArray(ctxStore, miner7.SectorsAmtBitwidth)
+	var inDeadlines miner6.Deadlines
+	err := store.Get(store.Context(), deadlines, &inDeadlines)
 	if err != nil {
-		return cid.Undef, xerrors.Errorf("failed to construct new sectors array: %w", err)
+		return cid.Undef, err
 	}
 
-	var sectorInfo miner6.SectorOnChainInfo
-	if err = inArray.ForEach(&sectorInfo, func(k int64) error {
-		return outArray.Set(uint64(k), migrateSectorInfo(sectorInfo))
-	}); err != nil {
-		return cid.Undef, err
+	var outDeadlines miner7.Deadlines
+	for i, c := range inDeadlines.Due {
+		if c == m.emptyDeadlineV6 && sectors == m.emptySectorsV7 {
```

Comment:
I'm not sure if this is correct. Usually, when we hit a "dead" deadline, we skip it and won't record any snapshots: https://github.com/filecoin-project/specs-actors/blob/d46a6510c2ad20ca453f72e292adbe880a8bacb8/actors/builtin/miner/miner_state.go#L1149. That means that never-used deadlines shouldn't have sector snapshots.

This is actually a really interesting edge-case because we can run into a state where we keep a _really old_ copy of the sectors array and never update it (if a deadline "dies"). This isn't a problem for _partitions_ because we only snapshot the deadline's partitions (so, when the deadline "dies", the partitions array and the partitions snapshot array are identical).